File: src/bin/read_when_write.cpp
#include <algorithm>
#include <iostream>

struct Data {
    int value;
    Data(int v) : value(v) {}
};

/**
 * @brief Expected to act like `out := max { 0, a, b }` in math.
 *
 */
void positive_max(Data &out, const Data &a, const Data &b) {
    out.value = 0;
    out.value = std::max(out.value, a.value);
    out.value = std::max(out.value, b.value);
}

void case1() {
    auto a = Data(1);
    const auto b = Data(-1);
    positive_max(a, b, a);
    std::cout << "a: " << a.value << std::endl;

    // Does not act like expected.

    // Output:
    // a: 0
}

void case2() {
    auto a = Data(1);
    auto b = Data(-1);
    auto c = Data(1);
    positive_max(a, b, c);
    std::cout << "a: " << a.value << std::endl;

    // Output:
    // a: 1
}

int main() {
    case1();
    case2();
    return 0;
}

File: src/bin/read_when_write.rs
struct Data {
    pub value: i32,
}
impl Data {
    pub fn new(value: i32) -> Self {
        Self { value }
    }
}

/**
 * Expected to act like `out := max { 0, a, b }` in math.
 */
fn positive_max(out: &mut Data, a: &Data, b: &Data) {
    out.value = 0;
    out.value = out.value.max(a.value);
    out.value = out.value.max(b.value);
}

// Compile Error:
// cannot borrow `a` as immutable because it is also borrowed as mutable immutable borrow occurs here
// fn case1() {
//     let mut a = Data::new(1);
//     let b = Data::new(-1);
//     positive_max(&mut a, &b, &a);
//     println!("a: {}", a.value);
// }

fn case2() {
    let mut a = Data::new(1);
    let b = Data::new(-1);
    let c = Data::new(1);
    positive_max(&mut a, &b, &c);
    println!("a: {}", a.value);

    // Output:
    // a: 1
}

fn main() {
    // case1();
    case2();
}
